File: mobile_client/modules/playaural-spatial-audio/src/PlayAuralSpatialAudioModule.ts
import { requireNativeModule } from "expo-modules-core";

export type SpatialAudioCapabilities = {
  available: boolean;
  hrtf: boolean;
  sampleRate: number;
};

export type SpatialAudioSourceOptions = {
  introPath?: string;
  loopPath: string;
  outroPath?: string;
  playIntro: boolean;
  looping: boolean;
  streamFromDisk: boolean;
  startPaused: boolean;
  volume: number;
  pitch: number;
  position: readonly [number, number, number];
  sequencePaths?: readonly string[];
  spatialBlend?: number;
};

export type SpatialAudioEngineOptions = {
  hrtfFrameSize: number;
  parameterSmoothingMilliseconds: number;
  maxSources: number;
};

export type PlayAuralSpatialAudioNativeModule = {
  initialize(options: SpatialAudioEngineOptions): Promise<SpatialAudioCapabilities>;
  createSource(sourceId: string, options: SpatialAudioSourceOptions): Promise<readonly number[]>;
  setParameters(
    sourceId: string,
    volume: number,
    pitch: number,
    x: number,
    y: number,
    z: number,
    spatialBlend: number,
  ): boolean;
  pauseSource(sourceId: string): boolean;
  resumeSource(sourceId: string): boolean;
  requestOutro(sourceId: string, finishLoopBoundary: boolean): boolean;
  destroySource(sourceId: string): void;
  drainEndedSources(): string[];
  shutdown(): void;
};

type PlayAuralSpatialAudioBridge = {
  initialize(
    hrtfFrameSize: number,
    parameterSmoothingMilliseconds: number,
    maxSources: number,
  ): Promise<SpatialAudioCapabilities>;
  createSource(sourceId: string, options: {
    introPath: string | null;
    loopPath: string;
    outroPath: string | null;
    playIntro: boolean;
    looping: boolean;
    streamFromDisk: boolean;
    startPaused: boolean;
    volume: number;
    pitch: number;
    x: number;
    y: number;
    z: number;
    spatialBlend: number;
    sequencePaths?: readonly string[];
  }): Promise<readonly number[]>;
  setParameters: PlayAuralSpatialAudioNativeModule["setParameters"];
  pauseSource: PlayAuralSpatialAudioNativeModule["pauseSource"];
  resumeSource: PlayAuralSpatialAudioNativeModule["resumeSource"];
  requestOutro: PlayAuralSpatialAudioNativeModule["requestOutro"];
  destroySource: PlayAuralSpatialAudioNativeModule["destroySource"];
  drainEndedSources: PlayAuralSpatialAudioNativeModule["drainEndedSources"];
  shutdown: PlayAuralSpatialAudioNativeModule["shutdown"];
};

const bridge = requireNativeModule<PlayAuralSpatialAudioBridge>(
  "PlayAuralSpatialAudio",
);

const module: PlayAuralSpatialAudioNativeModule = {
  initialize: (options) => bridge.initialize(
    options.hrtfFrameSize,
    options.parameterSmoothingMilliseconds,
    options.maxSources,
  ),
  createSource: (sourceId, options) => bridge.createSource(sourceId, {
    introPath: options.introPath ?? null,
    loopPath: options.loopPath,
    outroPath: options.outroPath ?? null,
    playIntro: options.playIntro,
    looping: options.looping,
    streamFromDisk: options.streamFromDisk,
    startPaused: options.startPaused,
    volume: options.volume,
    pitch: options.pitch,
    x: options.position[0],
    y: options.position[1],
    z: options.position[2],
    spatialBlend: options.spatialBlend ?? 1,
    ...(options.sequencePaths ? { sequencePaths: options.sequencePaths } : {}),
  }),
  setParameters: (...args) => bridge.setParameters(...args),
  pauseSource: (sourceId) => bridge.pauseSource(sourceId),
  resumeSource: (sourceId) => bridge.resumeSource(sourceId),
  requestOutro: (sourceId, boundary) => bridge.requestOutro(sourceId, boundary),
  destroySource: (sourceId) => bridge.destroySource(sourceId),
  drainEndedSources: () => bridge.drainEndedSources(),
  shutdown: () => bridge.shutdown(),
};

export default module;

File: mobile_client/modules/playaural-spatial-audio/android/src/main/java/one/ddt/playaural/spatialaudio/PlayAuralSpatialAudioModule.kt
package one.ddt.playaural.spatialaudio

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

private const val NATIVE_SUCCESS = 0
private const val MAX_SEQUENCE_SEGMENTS = 32
private val SOURCE_ID_PATTERN = Regex("^[A-Za-z0-9_.:-]{1,128}$")

private class SpatialAudioException(message: String) : CodedException(message)

private class SpatialAudioSourceOptions(
  @Field val introPath: String?,
  @Field val loopPath: String,
  @Field val outroPath: String?,
  @Field val playIntro: Boolean,
  @Field val looping: Boolean,
  @Field val streamFromDisk: Boolean,
  @Field val startPaused: Boolean,
  @Field val volume: Double,
  @Field val pitch: Double,
  @Field val x: Double,
  @Field val y: Double,
  @Field val z: Double,
  @Field val spatialBlend: Double,
  @Field val sequencePaths: List<String>?,
  @Field val sequenceNextStartRatios: List<Double>?
) : Record

class PlayAuralSpatialAudioModule : Module() {
  private val lock = Any()
  private val sources = LinkedHashMap<String, Long>()
  private var engineHandle = 0L
  private var engineConfiguration: Triple<Int, Int, Int>? = null

  override fun definition() = ModuleDefinition {
    Name("PlayAuralSpatialAudio")

    AsyncFunction("initialize") {
        hrtfFrameSize: Int,
        parameterSmoothingMilliseconds: Int,
        maxSources: Int ->
      synchronized(lock) {
        initializeEngine(hrtfFrameSize, parameterSmoothingMilliseconds, maxSources)
      }
    }

    AsyncFunction("createSource") { sourceId: String, options: SpatialAudioSourceOptions ->
      synchronized(lock) {
        createSource(sourceId, options)
      }
    }

    Function("setParameters") {
        sourceId: String,
        volume: Double,
        pitch: Double,
        x: Double,
        y: Double,
        z: Double,
        spatialBlend: Double ->
      synchronized(lock) {
        sources[sourceId]?.let { sourceHandle ->
          NativeSpatialAudioBridge.nativeSetParameters(
            sourceHandle,
            volume.toFloat(),
            pitch.toFloat(),
            x.toFloat(),
            y.toFloat(),
            z.toFloat(),
            spatialBlend.toFloat()
          ) == NATIVE_SUCCESS
        } ?: false
      }
    }

    Function("pauseSource") { sourceId: String ->
      synchronized(lock) {
        sources[sourceId]?.let {
          NativeSpatialAudioBridge.nativePauseSource(it) == NATIVE_SUCCESS
        } ?: false
      }
    }

    Function("resumeSource") { sourceId: String ->
      synchronized(lock) {
        sources[sourceId]?.let {
          NativeSpatialAudioBridge.nativeResumeSource(it) == NATIVE_SUCCESS
        } ?: false
      }
    }

    Function("requestOutro") { sourceId: String, finishLoopBoundary: Boolean ->
      synchronized(lock) {
        sources[sourceId]?.let {
          NativeSpatialAudioBridge.nativeRequestOutro(it, finishLoopBoundary) == NATIVE_SUCCESS
        } ?: false
      }
    }

    Function("destroySource") { sourceId: String ->
      synchronized(lock) {
        destroySource(sourceId)
      }
    }

    Function("drainEndedSources") {
      synchronized(lock) {
        val ended = mutableListOf<String>()
        val iterator = sources.iterator()
        while (iterator.hasNext()) {
          val entry = iterator.next()
          if (NativeSpatialAudioBridge.nativeSourceAtEnd(entry.value)) {
            NativeSpatialAudioBridge.nativeDestroySource(entry.value)
            iterator.remove()
            ended.add(entry.key)
          }
        }
        ended
      }
    }

    Function("shutdown") {
      synchronized(lock) {
        shutdown()
      }
    }

    OnDestroy {
      synchronized(lock) {
        shutdown()
      }
    }
  }

  private fun initializeEngine(
    hrtfFrameSize: Int,
    parameterSmoothingMilliseconds: Int,
    maxSources: Int
  ): Map<String, Any> {
    if (hrtfFrameSize <= 0 || parameterSmoothingMilliseconds < 0 || maxSources <= 0) {
      throw SpatialAudioException("Invalid native spatial-audio engine configuration")
    }
    val requested = Triple(hrtfFrameSize, parameterSmoothingMilliseconds, maxSources)
    if (engineHandle != 0L) {
      if (requested != engineConfiguration) {
        throw SpatialAudioException("Native spatial audio is already initialized with different settings")
      }
      return capabilities()
    }
    val result = NativeSpatialAudioBridge.nativeCreateEngine(
      hrtfFrameSize,
      parameterSmoothingMilliseconds,
      maxSources
    )
    if (result.size != 2 || result[0] == 0L || result[1] != NATIVE_SUCCESS.toLong()) {
      throw SpatialAudioException("Native spatial-audio initialization failed (${result.getOrNull(1)})")
    }
    engineHandle = result[0]
    engineConfiguration = requested
    return capabilities()
  }

  private fun capabilities(): Map<String, Any> = mapOf(
    "available" to (engineHandle != 0L),
    "hrtf" to (engineHandle != 0L),
    "sampleRate" to if (engineHandle != 0L) {
      NativeSpatialAudioBridge.nativeEngineSampleRate(engineHandle)
    } else {
      0
    }
  )

  private fun createSource(
    sourceId: String,
    options: SpatialAudioSourceOptions
  ): List<Double> {
    if (engineHandle == 0L) {
      throw SpatialAudioException("Native spatial audio is not initialized")
    }
    val sequencePaths = options.sequencePaths.orEmpty()
    val sequenceNextStartRatios = options.sequenceNextStartRatios.orEmpty()
    val hasStem = options.loopPath.isNotBlank()
    val hasSequence = sequencePaths.isNotEmpty()
    if (
      !SOURCE_ID_PATTERN.matches(sourceId) ||
      hasStem == hasSequence ||
      options.loopPath.contains('\u0000') ||
      options.introPath?.contains('\u0000') == true ||
      options.outroPath?.contains('\u0000') == true ||
      sequencePaths.size > MAX_SEQUENCE_SEGMENTS ||
      sequencePaths.any { it.isBlank() || it.contains('\u0000') } ||
      (hasSequence && sequenceNextStartRatios.size != sequencePaths.size) ||
      sequenceNextStartRatios.any { !it.isFinite() || it < 0.0 || it > 1.0 } ||
      (hasSequence && (
        !options.introPath.isNullOrEmpty() ||
        !options.outroPath.isNullOrEmpty() ||
        options.playIntro ||
        options.looping ||
        options.streamFromDisk
      ))
    ) {
      throw SpatialAudioException("Invalid native spatial-audio source")
    }
    destroySource(sourceId)
    val result = if (hasSequence) {
      NativeSpatialAudioBridge.nativeCreateSequenceSource(
        engineHandle,
        sequencePaths.toTypedArray(),
        sequenceNextStartRatios.map { it.toFloat() }.toFloatArray(),
        options.startPaused,
        options.volume.toFloat(),
        options.pitch.toFloat(),
        options.x.toFloat(),
        options.y.toFloat(),
        options.z.toFloat(),
        options.spatialBlend.toFloat()
      )
    } else {
      NativeSpatialAudioBridge.nativeCreateSource(
        engineHandle,
        options.introPath?.takeIf { it.isNotBlank() },
        options.loopPath,
        options.outroPath?.takeIf { it.isNotBlank() },
        options.playIntro,
        options.looping,
        options.streamFromDisk,
        options.startPaused,
        options.volume.toFloat(),
        options.pitch.toFloat(),
        options.x.toFloat(),
        options.y.toFloat(),
        options.z.toFloat(),
        options.spatialBlend.toFloat()
      )
    }
    if (result.size != 2 || result[0] == 0L || result[1] != NATIVE_SUCCESS.toLong()) {
      throw SpatialAudioException("Native spatial-audio source creation failed (${result.getOrNull(1)})")
    }
    val handle = result[0]
    val durations = if (hasSequence) {
      val sampleRate = NativeSpatialAudioBridge.nativeEngineSampleRate(engineHandle)
      val durationFrames = NativeSpatialAudioBridge.nativeSequenceDurations(handle)
      if (
        sampleRate <= 0 ||
        durationFrames.size != sequencePaths.size ||
        durationFrames.any { it <= 0 }
      ) {
        NativeSpatialAudioBridge.nativeDestroySource(handle)
        throw SpatialAudioException("Native spatial-audio sequence timing is invalid")
      }
      durationFrames.map { frames -> frames.toDouble() * 1000.0 / sampleRate }
    } else {
      emptyList()
    }
    sources[sourceId] = handle
    return durations
  }

  private fun destroySource(sourceId: String) {
    sources.remove(sourceId)?.let { sourceHandle ->
      NativeSpatialAudioBridge.nativeStopSource(sourceHandle)
      NativeSpatialAudioBridge.nativeDestroySource(sourceHandle)
    }
  }

  private fun shutdown() {
    sources.values.forEach { sourceHandle ->
      NativeSpatialAudioBridge.nativeStopSource(sourceHandle)
      NativeSpatialAudioBridge.nativeDestroySource(sourceHandle)
    }
    sources.clear()
    if (engineHandle != 0L) {
      NativeSpatialAudioBridge.nativeDestroyEngine(engineHandle)
      engineHandle = 0L
    }
    engineConfiguration = null
  }
}
